java/simplesaga-action/src/main/java/io/simplesource/saga/action/internal/AbstractActionTopologyBuilder.java

```diff
@@ -0,0 +1,76 @@
+package io.simplesource.saga.action.internal;
+
+import io.simplesource.saga.model.messages.ActionRequest;
+import io.simplesource.saga.model.messages.ActionResponse;
+import io.simplesource.saga.model.specs.ActionProcessorSpec;
+import io.simplesource.saga.shared.topics.TopicConfig;
+import io.simplesource.saga.shared.utils.StreamAppConfig;
+import lombok.Value;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.kstream.KStream;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+import java.util.UUID;
+
+public class AbstractActionTopologyBuilder<A> implements TopologyBuilder {
+
+    private final ActionProcessorSpec<A> actionSpec;
+    private final TopicConfig actionTopicConfig;
+    private List<TopologyBuildListener<A>> buildListeners = new ArrayList<>();
+
+    /**
+     * Topology being built.
+     */
+    @Value
+    public static class ActionTopologyContext<A> {
+        StreamsBuilder builder;
+        KStream<UUID, ActionRequest<A>> actionRequests;
+        KStream<UUID, ActionResponse> actionResponses;
+        Properties properties;
+    }
+
+    /**
+     * Called when the topology is built, ie. to allow sub-topologies to be added.
+     */
+    public interface TopologyBuildListener<A> {
```

Comment:
I'd be happy to eliminate this interface and replace it with `Consumer<ActionTopologyContext<A>>`.
Not a big deal though.

Reply:
Fixed